sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java

```diff
@@ -350,31 +363,31 @@ public void testUnboundedSourceCheckpointMark() throws Exception {
 
     UnboundedReader<KafkaRecord<byte[], Long>> reader = source.createReader(null, null);
     final int numToSkip = 3;
-    // advance once:
-    assertTrue(reader.start());
 
-    // Advance the source numToSkip-1 elements and manually save state.
-    for (long l = 0; l < numToSkip - 1; ++l) {
-      assertTrue(reader.advance());
+    // advance numToSkip elements
+    for (long l = 0; l < numToSkip; ++l) {
+      if (l > 0 || !reader.start()) {
+        advanceOnce(reader);
+      }
     }
 
     // Confirm that we get the expected element in sequence before checkpointing.
-
     assertEquals(numToSkip - 1, (long) reader.getCurrent().getKV().getValue());
     assertEquals(numToSkip - 1, reader.getCurrentTimestamp().getMillis());
 
     // Checkpoint and restart, and confirm that the source continues correctly.
     KafkaCheckpointMark mark = CoderUtils.clone(
         source.getCheckpointMarkCoder(), (KafkaCheckpointMark) reader.getCheckpointMark());
     reader = source.createReader(null, mark);
-    assertTrue(reader.start());
 
     // Confirm that we get the next elements in sequence.
     // This also confirms that Reader interleaves records from each partitions by the reader.
     for (int i = numToSkip; i < numElements; i++) {
+      if (i > numToSkip || !reader.start()) {
+        advanceOnce(reader);
+      }
```

Comment:
:). I was thinking of that too. Our start()/advance() interface is going to lead to many of these.